signal-hook-tokio/src/lib.rs

```diff
@@ -4,16 +4,57 @@
 
 //! A crate for integrating signal handling with the Tokio runtime.
 //!
-//! There are different sub modules for supporting different Tokio
-//! versions. The support for a version must be activated by a
-//! feature flag:
+//! This provides the [`Signals`][Signals] struct which acts as a
+//! [`Stream`][`futures::stream::Stream`] of signals.
```

Comment:
:thinking: I'm bit unsure here. So we are going to „stabilize“ in the form we are going to support one true version of tokio & friends. But:

* There's tokio-stream at 0.1.
* We are using futures, but 0.3, not 1.0 (which also doesn't exist AFAIK).

I wonder which one of these we *should* be using or if we should support multiple by feature flags (by calling into some kind of `poll_next_inner`).

Reply:
The lack of stable release for futures is an interesting point. Presumably the concern here is that the `Stream` trait changing in a future version. If that were to happen, or it were to be moved into the standard library as https://github.com/rust-lang/rfcs/pull/2996 proposes, it may be necessary to support multiple versions of `Stream` for a time. In the event that happens it seems reasonable to reintroduce feature flags to select which should be used. Do you see any need to do so proactively here?

 I don't see any direct dependency, direct or otherwise, on tokio-stream so I'm not sure clear what you're referring to there.

Reply:
Well, the concern is not exactly about _changing_. I don't really expect that trait to have a different shape, there's not much space for other design that the single `poll_next` method. But even if the trait looks exactly the same, if it comes from a different crate (or different version), it becomes a different trait. And removing support for some older version would be a breaking change. That's why I would like to support the features proactively, so user can remove the dependency on 0.3 eventually, while we don't yet drop support for it. But as there would be only one instance right now, it would probably mean just slapping some `#[cfg(feature = "futures_0_3")] onto the `impl Stream for ...` block, to make sure the users that need it enable it.

I've looked at `tokio-stream` in more detail and it just reexports the `futures`' stream, so we are good there. Nevertheless, I've noticed that the `Stream` actually comes from `futures-core`. I'd suggest we depend on that one, as it is smaller dependency than `futures`.

Reply:
Thanks for the explanation. I can see why it'd be useful to feature-gate it despite only a single version being relevant at present. I wasn't sure about the best way to achieve this. After some trial and error, I went with an approach that uses `cfg_if` to conditionally enable the `Stream` impl and supporting code while leaving the `Signals` type itself available unconditionally. Please let me know what you think.

I went ahead and switched the dependency on `futures` over to `futures-core` while leaving the references in the documentation as referring to `futures` for sake of familiarity for readers.